File: prolog/space-age/space_age.pl
space_age(Planet, AgeSec, Years) :-
    orbital_factor(Planet, Factor),
    Years is (AgeSec / 31557600) / Factor.

orbital_factor("Mercury", 0.2408467).
orbital_factor("Venus",   0.61519726).
orbital_factor("Earth",   1.0).
orbital_factor("Mars",    1.8808158).
orbital_factor("Jupiter", 11.862615).
orbital_factor("Saturn",  29.447498).
orbital_factor("Uranus",  84.016846).
orbital_factor("Neptune", 164.79132).

File: zig/space-age/space_age.zig
pub const Planet = enum {
    mercury,
    venus,
    earth,
    mars,
    jupiter,
    saturn,
    uranus,
    neptune,

    pub fn age(self: Planet, seconds: usize) f64 {
        var years: f64 = @intToFloat(f64, seconds) / 31557600.0;

        return years / self.factor();
    }

    fn factor(self: Planet) f64 {
        switch (self) {
            Planet.mercury => return 0.2408467,
            Planet.venus => return 0.61519726,
            Planet.earth => return 1.0,
            Planet.mars => return 1.8808158,
            Planet.jupiter => return 11.862615,
            Planet.saturn => return 29.447498,
            Planet.uranus => return 84.016846,
            Planet.neptune => return 164.79132,
        }
    }
};
